keps/sig-api-machinery/5589-gogo-dependency/README.md

```diff
@@ -0,0 +1,640 @@
+Kubernetes API Go instances or access fields.
+
+The changes in this proposal do not modify fields of Kubernetes API types.
+
+#### Story 3
+
+As a maintainer of Kubernetes code-generator components, I can merge changes to
+address vulnerabilities, bugs, and feature requirements in protobuf-handling.
+
+Once the protobuf generating code is forked from gogo/protobuf to k8s.io/code-generator
+or a sigs.k8s.io repo, it can be modified if needed to address vulnerabilities, bugs,
+and feature requirements.
+
+### Notes/Constraints/Caveats
+
+**Constraint: Compatibility with existing marshal / unmarshal behavior**
+
+Kubernetes API objects use protobuf wire encoding, but differ from typical protobuf behavior in two key ways:
+
+1. Canonical output for a given version is supported.
+   This is a requirement of Kubernetes API object serialization,
+   and is relied on by the API server storage layer to ensure repeated
+   no-op write requests do not re-persist different bytes to storage.
+   Canonical output is an [explicit non-goal](https://protobuf.dev/programming-guides/serialization-not-canonical/) of the standard protobuf libraries.
+
+2. Preservation of unknown fields is intentionally avoided.
+   Only fields which are known to the current API server version are allowed to be persisted and returned.
+   Preservation of unknown fields when decoding / round-tripping through the server is a non-goal for Kubernetes,
+   but is a non-configurable default behavior for `protoc`-generated marshaling / unmarshaling.
+
+These two key differences preclude switching to using `protoc` / `protoc-gen-go` generated marshaling / unmarshaling.
+
+**Constraint: Verifiable compatibility**
+
+Any changes in marshal / unmarshal implementation must be able to be verified to behave identically.
+Due to the complexity of protobuf marshaling / unmarshaling logic, any non-inspectable change would be difficult to prove behaves identically.
+
+**Constraint: Avoiding dangerous or disruptive changes to Kubernetes API types**
+
+Kubernetes API Go types are *widely* used by non-generated Go code to:
+* construct objects
+* read fields
+* write fields
+
+Changing to `protoc`/`protoc-gen-go`-generated objects would:
+1. change all fields to pointers, making every direct access of existing non-pointer fields risk a panic
+2. add additional unexported fields to objects to handle things like unknown field preservation and size caching
+3. add shallow-copy-unsafe fields to objects, making any code passing non-pointer instances no longer pass vet checks
+4. change the Go names of API fields containing initialisms (like `URL`)
+5. remove the ability to customize JSON tags for backwards compatibility
+
+These changes would impact non-protobuf usage of Kubernetes API types,
+and preclude switching to using `protoc` / `protoc-gen-go` generated marshaling / unmarshaling.
+
+### Risks and Mitigations
+
+**Risk: consumers of removed generated code will have to adjust usage.**
```

Comment:
Just a side note: Linters like [protogetter](https://golangci-lint.run/docs/linters/configuration/#protogetter) will also break CI on dependency upgrade, because they now treat the code as real protobuf compared to gogo. I don't think that this is any higher risk than the method changes below.

Reply:
> Linters like [protogetter](https://golangci-lint.run/docs/linters/configuration/#protogetter) will also break CI on dependency upgrade

I'm not sure that's accurate... there are no field getters generated today, so if a linter like that was unhappy, it would already be unhappy, right?

Reply:
ah, that's because _those_ go types actually *do* have getters now that they are generated by `protoc` / `protoc-gen-go`

Since we wouldn't be generating Kubernetes REST API types that way, they still would not have getters, so that wouldn't impact linters like that